docs/design_doc.md

```diff
@@ -0,0 +1,75 @@
+# Design document
+
+## Goals
+
+* Redesign the Github [mirror](#mirror) to reduce the number of Github API requests required to build up the mirror cache with every deployment. 
+
+* Cache rebuilding time after redeployment should be constant and not grow linearly in the number of [mirror replicas](#mirror-replica). 
+
+* Minimize latency so that the average mirror [request time](#request-time) does not increase significantly.
+
+## Current solution
+
+Each replica of the github mirror server maintains its own [in-memory cache](#in-memory-cache). The cache is a data structure (Python dictionary) with constant-time lookup. The cache may grow unbounded if the pod does not fail or get redeployed.
+
+The current design incurs a significant overhead during redeployment as each replica has to build and maintain its own copy of the cache. This design does not scale well if the number of replicas is increased, as the [hourly quota](https://developer.github.com/v3/#rate-limiting) for Github API requests will quickly be exhausted, causing a hold up for all [clients](#client) depending on this service.
+
+## Proposed solution
+
+* Replace each pod's in-memory cache with a single shared cache for all pods. 
+
+    * The [cache server](#cache-server) should be deployed within the same Open Shift cluster as the Github mirror to reduce the latency associated with requests to the cache.
+
+    * A shared cache should reduce the number of calls to the Github API since each pod will not maintain its own copy of the cache.
+
+    * When the mirror is redeployed, the cache will have to be rebuilt once only, instead of once for *each* replica.
+
+* A highly available key-value store such as Redis can be used as the cache server.
+
+    * The [etags](https://developer.github.com/v3/#conditional-requests) returned with each conditional request to the Github API are unique 
+
+    * Redis does not guarantee [strong consistency](https://redis.io/topics/cluster-tutorial#redis-cluster-consistency-guarantees) which is acceptable since any stale cache reads will always be compared against the response from the conditional request, resulting - in the worst case - in an additional request to the Github api as well as unnecessairly replacing the cache entry.
+
+This redesign should enable multi-processing for the mirror since different processes can access the shared cache server which is currently not possible when using an in-memory data structure.
+
+### Architecture diagram
+
+![](./images/arch.png)
```

Comment:
could you please commit the source of this image? that's useful for when somone has to extend the diagrams.
also, the multiple arrows from the client to the github-mirror service gives the impression that the load balancing is achieved in the client side, which is not true. A single arrow to the outer box will do it.

Reply:
good point. will do.